### PR TITLE
ARROW-7477: [Java][FlightRPC] set up gRPC reflection metadata

### PR DESCRIPTION
**Pull request by lidavidm**
This sets up the gRPC/Protobuf reflection metadata, so that [gRPC reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md) works. This is useful for debugging. Additionally, it means interceptors can do generic introspection on gRPC and Flight methods, without having to special-case them. (You might want to do this to implement path-based access control, for instance.)

### CODE REVIEW

File: java/flight/src/main/java/org/apache/arrow/flight/FlightBindingService.java [hunk 1]
```diff
@@ -82,7 +86,21 @@ public ServerServiceDefinition bindService() {
...
     final MethodDescriptor<ArrowMessage, Flight.PutResult> doPutDescriptor = getDoPutDescriptor(allocator);
 
-    ServerServiceDefinition.Builder serviceBuilder = ServerServiceDefinition.builder(FlightConstants.SERVICE);
+    // Make sure we preserve SchemaDescriptor fields on methods so that gRPC reflection still works.
+    final ServiceDescriptor.Builder serviceDescriptorBuilder = ServiceDescriptor.newBuilder(FlightConstants.SERVICE)
+        .setSchemaDescriptor(baseDefinition.getServiceDescriptor().getSchemaDescriptor());
+    serviceDescriptorBuilder.addMethod(doGetDescriptor);
+    serviceDescriptorBuilder.addMethod(doPutDescriptor);
+    for (MethodDescriptor<?, ?> definition : baseDefinition.getServiceDescriptor().getMethods()) {
+      if (OVERRIDE_METHODS.contains(definition.getFullMethodName())) {
```
Comment (liyafan82) [inline java/flight/src/main/java/org/apache/arrow/flight/FlightBindingService.java]: I do not have much experience in gRPC. Are method names case-sensitive?

File: java/flight/src/test/java/org/apache/arrow/flight/TestServerOptions.java [hunk 2]
```diff
@@ -128,4 +132,20 @@ public void domainSocket() throws Exception {
...
+  public void checkReflectionMetadata() {
+    // This metadata is needed for gRPC reflection to work.
+    final ExecutorService executorService = Executors.newSingleThreadExecutor();
+    try (final BufferAllocator allocator = new RootAllocator(Integer.MAX_VALUE)) {
+      final FlightBindingService service = new FlightBindingService(allocator, new NoOpFlightProducer(),
+          ServerAuthHandler.NO_OP, executorService);
+      final ServerServiceDefinition definition = service.bindService();
+      assertNotNull(definition.getServiceDescriptor().getSchemaDescriptor());
+      definition.getMethods().forEach(method -> assertNotNull(method.getMethodDescriptor().getSchemaDescriptor()));
+      definition.getServiceDescriptor().getMethods().forEach(method -> assertNotNull(method.getSchemaDescriptor()));
```
Comment (liyafan82) [inline java/flight/src/test/java/org/apache/arrow/flight/TestServerOptions.java]: Is it possible to do stronger validation here?

Comment (lidavidm) [reply java/flight/src/test/java/org/apache/arrow/flight/TestServerOptions.java]: I've reworked the test to both make sure that all methods are present, and that the reflection objects are equal between both places where they are accessible and equal to the original object.

File: java/flight/src/main/java/org/apache/arrow/flight/FlightBindingService.java (hunk 1 above)
Comment (lidavidm) [reply java/flight/src/main/java/org/apache/arrow/flight/FlightBindingService.java]: Yes, method names are case-sensitive. There should be no reason why the method name ever differs here, though.

**Review by jacques-n — APPROVED** [review]
+1 LGTM

File: java/flight/src/main/java/org/apache/arrow/flight/FlightBindingService.java (hunk 1 above)
Comment (liyafan82) [reply java/flight/src/main/java/org/apache/arrow/flight/FlightBindingService.java]: Thanks for the clarification.

**Review by liyafan82 — APPROVED** [review]

**Pull request closed without merging**